src/main/java/com/amazon/opendistroforelasticsearch/sql/parser/subquery/InSubqueryRewriter.java

```diff
@@ -0,0 +1,112 @@
+/*
+ *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.parser.subquery;
+
+import com.alibaba.druid.sql.ast.SQLExpr;
+import com.alibaba.druid.sql.ast.expr.SQLBinaryOpExpr;
+import com.alibaba.druid.sql.ast.expr.SQLBinaryOperator;
+import com.alibaba.druid.sql.ast.expr.SQLInSubQueryExpr;
+import com.alibaba.druid.sql.ast.expr.SQLNullExpr;
+import com.alibaba.druid.sql.ast.statement.SQLJoinTableSource;
+import com.alibaba.druid.sql.ast.statement.SQLJoinTableSource.JoinType;
+import com.alibaba.druid.sql.dialect.mysql.ast.statement.MySqlSelectQueryBlock;
+import com.amazon.opendistroforelasticsearch.sql.parser.subquery.rewriter.SubqueryAliasRewriter;
+
+/**
+ * IN Subquery Rewriter.
+ * For example,
+ *  SELECT * FROM A WHERE a IN (SELECT b FROM B) should be rewritten to
+ *  SELECT A.* FROM A JOIN B ON A.a = B.b.
+ */
+public class InSubqueryRewriter implements SubqueryRewriter {
+
+    private final MySqlSelectQueryBlock rootQuery;
+
+    private final SQLInSubQueryExpr inSubQueryExpr;
+
+    private final MySqlSelectQueryBlock inSubQuery;
+
+    private SQLExpr joinLeftExpr;
+
+    private SQLExpr joinRightExpr;
+
+    public InSubqueryRewriter(MySqlSelectQueryBlock rootQuery, SQLInSubQueryExpr inSubQueryExpr) {
+        rootQuery.accept0(new SubqueryAliasRewriter());
```

Comment:
can we move that rewrite operation into rewriteSubquery? It doesn't seems like there is anything in other parameters that depends on the rootQuery?

Reply:
Agree. It is a generic opeartion for all the subquery clause. Move it to SubqueryRewriterHelper.